### PR TITLE
drivers: wifi: Fix OSAL issues for other OSes

### PR DESCRIPTION
**Pull request by ajayparida**
Handle os specific code.
(1) Moved os specific code to shim to handle it os independent way. (2) Re-arranged code to build it for other os.

### CODE REVIEW

**Review by krish2718 — COMMENTED** [review]
Can you please split this in to smaller commits? And also, the commit log (both title and body) needs some work, its confusing .

File: drivers/wifi/nrf700x/zephyr/src/shim.c [hunk 1]
```diff
@@ -814,6 +832,10 @@ static const struct wifi_nrf_osal_ops wifi_nrf_os_zep_ops = {
 	.bus_qspi_ps_wake = zep_shim_bus_qspi_ps_wake,
 	.bus_qspi_ps_status = zep_shim_bus_qspi_ps_status,
 #endif /* CONFIG_NRF_WIFI_LOW_POWER */
+
+	.assert = zep_shim_assert,
+	.mem_cmp = zep_shim_mem_cmp,
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/shim.c:837]: Club this with other mem_ ops.

File: drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c [hunk 2]
```diff
@@ -36,8 +36,9 @@ static enum wifi_nrf_status hal_fw_patch_load(struct wifi_nrf_hal_dev_ctx *hal_d
 	int last_chunk_size = fw_patch_size % MAX_PATCH_CHUNK_SIZE;
 	int num_chunks = fw_patch_size / MAX_PATCH_CHUNK_SIZE +
 					(last_chunk_size ? 1 : 0);
+	int chunk = 0;
 
-	for (int chunk = 0; chunk < num_chunks; chunk++) {
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c:40]: unrelated to any of the other changes in this PR...should be a separate commit.

File: drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c [hunk 3]
```diff
@@ -160,7 +162,7 @@ enum wifi_nrf_status wifi_nrf_hal_fw_patch_load(struct wifi_nrf_hal_dev_ctx *hal
 		{ "bin", fw_sec_patch_data, fw_sec_patch_size, sec_dest_addr },
 	};
 
-	for (int patch = 0; patch < ARRAY_SIZE(patches); patch++) {
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c:163]: same comment as above...regarding change unrelated to other changes in the PR

**Review by sachinthegreen — CHANGES_REQUESTED** [review]
Please also split this PR into separate commits for assert, strlen and memcmp

**Comment by sachinthegreen** [comment]
> Please also split this PR into separate commits for assert, strlen and memcmp

Sorry looks like @krish2718 has already suggested this above.

**Comment by ajayparida** [comment]
> Can you please split this in to smaller commits? And also, the commit log (both title and body) needs some work, its confusing .

Done , Please review

**Comment by ajayparida** [comment]
> Please also split this PR into separate commits for assert, strlen and memcmp

Done, Please review.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/event.c [hunk 4]
```diff
@@ -17,7 +17,6 @@
 #include "fmac_peer.h"
 #include "fmac_cmd.h"
 #include "fmac_ap.h"
-#include <string.h>
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/event.c:20]: Commit log: `drivers: wifi: Remove unused header file`

File: drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c [hunk 5]
```diff
@@ -36,8 +36,9 @@ static enum wifi_nrf_status hal_fw_patch_load(struct wifi_nrf_hal_dev_ctx *hal_d
 	int last_chunk_size = fw_patch_size % MAX_PATCH_CHUNK_SIZE;
 	int num_chunks = fw_patch_size / MAX_PATCH_CHUNK_SIZE +
 					(last_chunk_size ? 1 : 0);
+	int chunk = 0;
 
-	for (int chunk = 0; chunk < num_chunks; chunk++) {
+	for (chunk = 0; chunk < num_chunks; chunk++) {
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/hw_if/hal/src/hal_fw_patch_loader.c:41]: Commit log:

```
drivers: wifi: Fix the code to work with any C standard

When using OSAL with Linux, some kernels are built with "gnu11 C" standard,
where declaring inline isn't allowed, but its allowed in C99, so, use a commonly
accepted explicit declaration.
```

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c [hunk 6]
```diff
@@ -29,7 +28,10 @@ static void set_spare_desc_q_map(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx,
 {
 	unsigned short spare_desc_indx = 0;
 
-	__ASSERT(fmac_dev_ctx->fpriv->num_tx_tokens_per_ac != 0, "num_tx_tokens_per_ac is zero");
+	wifi_nrf_osal_assert(fmac_dev_ctx->fpriv->opriv,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:31]: This isn't introduced in this commit.

Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:31]: Commit log:

```
drivers: wifi:  Add OS agnostic assertion support

This abstracts assertions to be used in OSAL code rather than
directly invoking OS specific assertions.
```

File: drivers/wifi/nrf700x/zephyr/src/shim.c [hunk 7]
```diff
@@ -822,6 +827,7 @@ static const struct wifi_nrf_osal_ops wifi_nrf_os_zep_ops = {
 #endif /* CONFIG_NRF_WIFI_LOW_POWER */
 
 	.assert = zep_shim_assert,
+	.str_len = zep_shim_str_len,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/shim.c:838]: `.strlen`

File: drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h [hunk 8]
```diff
@@ -1319,4 +1319,17 @@ void wifi_nrf_osal_assert(struct wifi_nrf_osal_priv *opriv,
...
+ * wifi_nrf_osal_str_len() - Gives the length of the string @str.
+ * @opriv: Pointer to the OSAL context returned by the @wifi_nrf_osal_init API.
+ * @str: Pointer to the memory location of the string.
+ *
+ * Calculates the length of the string pointed to by @str
+ *
+ * Return:
+ *              returns the number of bytes of the string @str.
+ */
+unsigned int wifi_nrf_osal_str_len(struct wifi_nrf_osal_priv *opriv,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:1333]: `strlen`

Comment (krish2718) [inline drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:1333]: commit log:

```
drivers: wifi: Add strlen to the OSAL ops

<string.h> header file location is different between OSes, so,
add a OSAL op to hide this from OS agnostic code.
```

Comment (krish2718) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h:1333]: If it's just a relative path, why can't we handle it by adding the path to includes in `CMakeLists.txt/Makefile` ?

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api_common.c [hunk 9]
```diff
@@ -411,7 +411,8 @@ enum wifi_nrf_status wifi_nrf_fmac_otp_mac_addr_get(struct wifi_nrf_fmac_dev_ctx
 				      otp_mac_addr,
 				      NRF_WIFI_ETH_ADDR_LEN);
 
-		if (!nrf_wifi_utils_is_mac_addr_valid((const char *)mac_addr)) {
+		if (!nrf_wifi_utils_is_mac_addr_valid(fmac_dev_ctx->fpriv->opriv,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api_common.c:414]: Use the same suggested commit log from earlier commit. And also same argument for `strlen` holds true here.

**Review by krish2718 — CHANGES_REQUESTED** [review]

**Review by sachinthegreen — APPROVED** [review]

File: drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h (hunk 8 above)
Comment (ajayparida) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h]: Commit log updated.
Inclusion of path in Makefile will not resolve inclusion of path to string.h header file in lnux case leads to other build issues.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api_common.c (hunk 9 above)
Comment (ajayparida) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api_common.c:414]: Commit log updated. 
Inclusion of Path in Makefile leads to oher build issue.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c (hunk 6 above)
Comment (ajayparida) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:31]: I think In my second commit I had updated. 
Now it's building for all the commit.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api_common.c (hunk 9 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api_common.c:414]: Can you please share the issue? We should be able to fix this easily by appending to the include path.

File: drivers/wifi/nrf700x/osal/os_if/src/osal.c [hunk 10]
```diff
@@ -756,3 +756,11 @@ int wifi_nrf_osal_bus_qspi_ps_status(struct wifi_nrf_osal_priv *opriv,
 	return opriv->ops->bus_qspi_ps_status(os_qspi_priv);
 }
 #endif /* CONFIG_NRF_WIFI_LOW_POWER */
+
+void wifi_nrf_osal_assert(struct wifi_nrf_osal_priv *opriv,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/os_if/src/osal.c:771]: Why dont' we use a generic macro directly defined by the shim? This is is unnecessary and also isn't giving flexibility is using other conditions than `!=`. 
E.g.,

In Zephyr:

#define WIFI_NRF_OSAL_ASSERT ASSERT()

In Linux:

#define WIFI_NRF_OSAL_ASSERT BUG_ON()

File: drivers/wifi/nrf700x/zephyr/src/shim.c [hunk 11]
```diff
@@ -733,6 +734,11 @@ static void zep_shim_timer_kill(void *timer)
 }
 #endif /* CONFIG_NRF_WIFI_LOW_POWER */
 
+static void zep_shim_assert(int test_val, int val, char *msg)
+{
+	__ASSERT(test_val != val, "%s", msg);
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/shim.c:746]: networking modules/drivers should use `NET_ASSERT`

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/osal/os_if/src/osal.c (hunk 10 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/os_if/src/osal.c:771]: Well, this won't work directly as we can't include SHIM headers in OSAL code.

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c [hunk 12]
```diff
@@ -49,7 +52,11 @@ static void clear_spare_desc_q_map(struct wifi_nrf_fmac_dev_ctx *fmac_dev_ctx,
...
+			     fmac_dev_ctx->fpriv->num_tx_tokens_per_ac,
+			     0,
+			     WIFI_NRF_ASEERT_NOT_EQUAL_TO,
```
Comment (krish2718) [inline drivers/wifi/nrf700x/osal/fw_if/umac_if/src/tx.c:58]: ASSERT

File: drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api_common.c (hunk 9 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/fw_if/umac_if/src/fmac_api_common.c:414]: Offline discussion:

The same C headers are defined in Linux kernel as well, so, if we fix the include path to include `string.h` the automatically the `stddef.h` from kernel is picked up which doesn't have `size_t` defined and causes build issues.

Also, using C standard functions may not be enough, as Linux kernel leverages processor architecture specific features to re-implement C standard functions efficiently, e.g., `memcmp`, so, its better abstract these and let shim implement as per the OS.

**Review by krish2718 — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h (hunk 8 above)
Comment (krish2718) [reply drivers/wifi/nrf700x/osal/os_if/inc/osal_api.h]: Same as https://github.com/nrfconnect/sdk-nrf/pull/12229#discussion_r1319747744

**Review by krish2718 — APPROVED** [review]
